nexus/db-queries/src/db/datastore/test_utils.rs

```diff
@@ -0,0 +1,343 @@
+
+        let expunged_fut = async {
+            sled_set_policy(
+                &opctx,
+                &datastore,
+                self.expunged,
+                SledPolicy::Expunged,
+                ValidateTransition::Yes,
+                Expected::Ok(SledPolicy::provisionable()),
+            )
+            .await
+            .with_context(|| {
+                format!(
+                    "failed to set expunged policy for sled {}",
+                    self.non_provisionable
+                )
+            })
+        };
+
+        // Legally, we must set the policy to expunged before setting the state
+        // to decommissioned. (In the future, we'll want to test graceful
+        // removal as well.)
+        let decommissioned_fut = async {
+            sled_set_policy(
+                &opctx,
+                &datastore,
+                self.decommissioned,
+                SledPolicy::Expunged,
+                ValidateTransition::Yes,
+                Expected::Ok(SledPolicy::provisionable()),
+            )
+            .await
+            .with_context(|| {
+                format!(
+                    "failed to set expunged policy for sled {}, \
+                     as prerequisite for decommissioning",
+                    self.decommissioned
+                )
+            })?;
+
+            sled_set_state(
+                &opctx,
+                &datastore,
+                self.decommissioned,
+                SledState::Decommissioned,
+                ValidateTransition::Yes,
+                Expected::Ok(SledState::Active),
+            )
+            .await
+            .with_context(|| {
+                format!(
+                    "failed to set decommissioned state for sled {}",
+                    self.decommissioned
+                )
+            })
+        };
+
+        // This is _not_ a legal state, BUT we test it out to ensure that if
+        // the system somehow enters this state anyway, we don't try and
+        // provision resources on it.
+        let illegal_decommissioned_fut = async {
+            sled_set_state(
+                &opctx,
+                &datastore,
+                self.illegal_decommissioned,
+                SledState::Decommissioned,
+                ValidateTransition::No,
+                Expected::Ok(SledState::Active),
+            )
+            .await
+            .with_context(|| {
+                format!(
+                    "failed to illegally set decommissioned state for sled {}",
+                    self.illegal_decommissioned
+                )
+            })
+        };
+
+        // We're okay cancelling the rest of the futures if one of them fails,
+        // since the overall test is going to fail anyway. Hence try_join
+        // rather than join_then_try.
+        futures::try_join!(
+            non_provisionable_fut,
+            expunged_fut,
+            decommissioned_fut,
+            illegal_decommissioned_fut
+        )?;
```

Comment:
I do tend to prefer using anyhow errors in tests, because it lets you write subtests with good error messages, like in `test_sled_transitions`. (I've learned this the hard way over time.)

`RUST_BACKTRACE=1` shows the point at which the error was raised.